botorch/acquisition/knowledge_gradient.py

```diff
@@ -46,12 +47,456 @@
+        )
+        noise_variance = self.model.likelihood.noise_covar.noise
+        full_posterior_mean = full_posterior.mean  # (1 + num_X_disc , 1)
+
+        # Compute full Covariante Cov(Xnew, X_discretised), select [Xnew X_discretised] submatrix, and subvectors.
+        discretisation_mean = full_posterior_mean[
+            len(xnew) :
+        ].squeeze()  # (num_X_disc, )
+        full_posterior_covariance = (
+            full_posterior.mvn.covariance_matrix
+        )  # (1 + num_X_disc , 1 + num_X_disc )
+        posterior_cov_xnew_opt_disc = full_posterior_covariance[
+            : len(xnew), :
+        ].squeeze()  # ( 1 + num_X_disc,)
+        full_posterior_variance = (
+            full_posterior.variance.squeeze()
+        )  # (1 + num_X_disc, )
+        full_predictive_covariance = (
+            posterior_cov_xnew_opt_disc
+            / (full_posterior_variance + noise_variance).sqrt()
+        )
+        # initialise empty kgvals torch.tensor
+        kgval = self.kgcb(a=full_posterior_mean, b=full_predictive_covariance)
+
+        return kgval
+
+    def compute_mc_kg(self, xnew: Tensor, Zvals: Tensor) -> tuple[Tensor, Tensor]:
+        """
+
+        Args:
+            xnew: A `1 x 1 x d` Tensor with `1` acquisition function evaluations of
+            `d` dimensions.
+            Zvals: 1 x num_fantasies Tensor with num_fantasies Normal quantiles.
+
+        Returns:
+            xstar_inner_optimisation: num_fantasies x d Tensor. Optimal X values for each z in Zvals.
+            xnew_acquisition_value: 1 x 1 Tensor, Monte Carlo KG value of xnew
+        """
+
+        # There's a recurssion problem importing packages by one_shot kg. Hacky way of importing these packages.
+        # TODO: find a better way to fix this
+        if "gen_candidates_scipy" not in sys.modules:
+            from botorch.generation.gen import gen_candidates_scipy
+            from botorch.optim.initializers import gen_value_function_initial_conditions
+
+        expected_xnew_value = torch.zeros(
+            xnew.shape[0], dtype=torch.double
+        )  # (m x 1 x d)
+        # Loop over xnew points
+        fantasy_opt_val = torch.zeros((1, self.num_fantasies))  # 1 x num_fantasies
+        xstar_inner_optimisation = torch.zeros((self.num_fantasies, xnew.shape[1]))
+
+        # This setting makes sure that I can rewrite the base samples and use the quantiles.
+        # Therefore, resample=False, collapse_batch_dims=True.
+        sampler = SobolQMCNormalSampler(
+            num_samples=1, resample=False, collapse_batch_dims=True
+        )
+
+        # loop over number of GP fantasised mean realisations
+        for fantasy_idx in range(self.num_fantasies):
+            # construct one realisation of the fantasy model by adding xnew. We rewrite the internal variable
+            # base samples, such that the samples are taken from the quantile.
+            zval = Zvals[fantasy_idx].view(1, 1, 1)
```

Comment:
add a TODO comment (don't need to make changes right now)
```python
# TODO: we have zvals in as a function arguments, we do not need to sample them
# from the SobolQMCsampler. However `fantasize()` requirtes a sampler argument sooo......
```

Reply:
done